x/stake/keeper/validator.go

```diff
@@ -246,13 +257,69 @@ func (k Keeper) UpdateValidator(ctx sdk.Context, validator types.Validator) type
 			bz := k.cdc.MustMarshalBinary(validator.ABCIValidator())
 			store.Set(GetTendermintUpdatesKey(validator.Owner), bz)
 		}
-
 	}
 
 	k.SetValidator(ctx, validator)
 	return validator
 }
 
+// updateCliffValidator determines if the current cliff validator needs to be
+// updated or swapped. If the provided affected validator is the current cliff
+// validator before it's power was increased, either the cliff power key will
+// be updated or if it's power is greater than the next bonded validator by
+// power, it'll be swapped.
+func (k Keeper) updateCliffValidator(ctx sdk.Context, affectedVal types.Validator) {
+	var newCliffVal types.Validator
+
+	store := ctx.KVStore(k.storeKey)
+	pool := k.GetPool(ctx)
+	cliffAddr := sdk.AccAddress(k.GetCliffValidator(ctx))
+
+	oldCliffVal, found := k.GetValidator(ctx, cliffAddr)
+	if !found {
+		panic(fmt.Sprintf("cliff validator record not found for address: %v\n", cliffAddr))
+	}
+
+	// NOTE: We get the power via affectedVal since the store (by power key)
+	// has yet to be updated.
+	affectedValPower := affectedVal.GetPower()
+
+	// Create a validator iterator ranging from smallest to largest by power
+	// starting the current cliff validator's power.
+	start := GetValidatorsByPowerIndexKey(oldCliffVal, pool)
+	end := sdk.PrefixEndBytes(ValidatorsByPowerIndexKey)
+	iterator := store.Iterator(start, end)
+
+	if iterator.Valid() {
+		ownerAddr := iterator.Value()
+		currVal, found := k.GetValidator(ctx, ownerAddr)
+		if !found {
+			panic(fmt.Sprintf("validator record not found for address: %v\n", ownerAddr))
+		}
+
+		if currVal.Status != sdk.Bonded || currVal.Revoked {
+			panic(fmt.Sprintf("unexpected revoked or unbonded validator for address: %s\n", ownerAddr))
+		}
+
+		newCliffVal = currVal
+		iterator.Close()
+	} else {
+		panic("failed to create valid validator power iterator")
```

Comment:
I wonder if this might crash if the user is running only one validator. Not sure if that is a usecase we care about or not - but we could check it with `params.MaxVals`

Reply:
We probably want to support that for convenience in testing.

Reply:
This should not crash, but let me check.

Reply:
Yeah it works with one validator 👍

Reply:
Sweet! Lets merge this then!